### ISSUE TITLE
card.html full of errors

### ISSUE THREAD

**Issue by Fred-Vatin** [ankidroid/Anki-Android#5273]
### Reproduction Steps

1. Look at this **card.html** on [Codepen](https://codepen.io/freMea/pen/NJVwej?editors=1100) using a Chrome based browser. The HTML is an untouched (except the image address) **card.html** from AnkiDroid.

2. Paste the html in [validator.w3.org](https://validator.w3.org/#validate_by_input)
---
### Actual Result

Lots of errors. 

One of them is `#qa` is not accessible. 

```html
<div class="answer id=" qa ">
```
It should be  
```html
<div class="answer" id="qa">
```

And if you try to style your card using `#qa`, it won’t work on AnkiDroid but will in desktop.

---
Another is AnkiDroid add whitespace in ID attribute and about everywhere after or before some chars `( " `.

Here is a piece of an original code I edited in Anki:
```html
<div class="front front-back">
	<div class="number">0</div>
	<a class="show_hint" href="#" title="Show hint" onclick="this.style.display = 'none';var el=document.getElementById('hint');el.style.display='inline-block';return false;"></a>
	<div id="hint" style="display: none">Personnage de film</div>
</div>
```

and here is the same once saved in card.html by AnkiDroid:
```html
<a class="show_hint " href="# " title="Show hint " onclick="this.style.display='none' ;var el=document.getElementById( 'hint');el.classList.add( 'hint');el.style.display='inline-block' ;return false; "></a>
<div id="hint " style="display: none ">Personnage de film</div>
```

The whitespace in `id="hint "` should break JS code using `document.getElementById('hint')`.

And it does if you test this in Codepen. Pressing the _show hint_ button produces this error in console:

```
index.html?editors=1100:300 Uncaught TypeError: Cannot read property 'style' of null
    at HTMLAnchorElement.onclick (index.html?editors=1100:300)
onclick @ index.html?editors=1100:300

```
**But weirdly**, the button works in AnkiDroid. Hence, I wonder if this whitespace issue is not related to the **card.html** saving process and if this HTML is exactly the same AnkiDroid use to render the card.

---
### Expected Result

- No such issue on Anki destop & web version. `#qa` is accessible for CSS styling.
- A produced card.html error proof that doesn’t add whitespace in attributes, JS user script, etc.

---
### Debug info

```
AnkiDroid Version = 2.9alpha67

Android Version = 9

ACRA UUID = cce85e92-250e-4aa0-9955-4e495676f57d
```

---
### Research

[ x ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[ x ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[ x ] I have searched for similar existing issues here and on the user forum

**Comment by mikehardy** [comment]
Very interesting - nice work digging in with the alphas. It looks like you based your research and fix on the generated card.html - now that is the string that we should send in to the webview directly so there shouldn't be a difference between what went on to the filesystem and what went into the webview, but did you try with the actual devtools hooked up to the webview live? Just curious - as that would let us know if it was the file save or if we just corrupted the attributes somehow.

Tracking this down I see:
https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/assets/card_template.html

is populated by 

https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java#L2153

and that content variable is passed through lots of filters... If you have any of the Android dev tools set up (Android Studio and at least one emulator - fairly easy to set up though they are huge downloads) you could Timber.d(content) as it goes in and out of things to find out where your things were being munged

If you don't have those set up this will have to sit for a little while as I don't have time to chase it right now, sorry

**Comment by freMea** [comment]
> did you try with the actual devtools hooked up to the webview live?

I regret I don’t understand how to do that. I’m not a android dev either. I can’t do much about it than report bugs. Sorry.

The good news is I managed to style the output as I want. The user just has to know that it can not use `#qa` as he does with Anki desktop.

**Comment by mikehardy** [comment]
Glad you were able to get your cards styled - can you share how you selected it via CSS if it wasn't via `#qa` ?

**Comment by freMea** [comment]
> 

> Glad you were able to get your cards styled - can you share how you selected it via CSS if it wasn't via `#qa` ?

Via `.answer, .question`

**Comment by ohare93** [comment]
Perhaps this is related. I recently made a new styling for my cards

```
.answer {
 color:yellow;
}
```

which I referenced with

`<span class=answer>{{X Word}}</span>`

On the back of my card template. What happened instead of this part being yellow was that the whole card turned yellow whenever I flipped to the back... this would not happen on Anki desktop or on iOS on a tablet, only on AnkiDroid. Simply renaming ".answer" to anything else fixed the issue. Seems like "answer" is a special field name in AnkiDroid, whereas it is not in other versions.

**Comment by freMea** [comment]
> Perhaps this is related. I recently made a new styling for my cards
> 
>  ```
> .answer {
>  color:yellow;
> }
> ```
> which I referenced with
> 
> `<span class=answer>{{X Word}}</span>`

You can’t use `.answer` alone since it already exists in the DOM at a higher level. Change the class name or use `span.answer`

**Comment by ohare93** [comment]
When you say "can't use" do you mean "shouldn't"? If so, sure thing 👍 But on Anki desktop and iOS it works perfectly fine with "answer" as a style. Everything works, colours, size, font, etc all work exactly where you set them. Only on the android version does style get applied to the Back of all cards (i.e. the "Answer"). It is an inconsistency.

**Comment by mikehardy** [comment]
inconsistencies with anki ecosystem are (in my opinion) always bugs to be fixed

**Comment by mikehardy** [comment]
I had a separate issue somewhere (can't find the link now?) where the deck templates we during initialization have gone stale with regard to the ones that Anki Desktop uses, and that may be related - looking upstream at the default templates and getting rid of the skew may solve both items at once

**Comment by milyukhin** [comment]
This is partially fixed in #5356. Except the `The whitespace in id="hint " should break JS` part (need to be verified @ohare93 can you please add steps how to reproduce it?)

**Comment by BrayanDSO** [comment]
Closing in favor of #14302
